src/api.cc

```diff
@@ -119,57 +127,91 @@ AkliteClient::~AkliteClient() {
-  for (const auto& t : client_->allTargets()) {
+  for (const auto& t : allTargets) {
     int ver = 0;
     try {
       ver = std::stoi(t.custom_version(), nullptr, 0);
     } catch (const std::invalid_argument& exc) {
       LOG_ERROR << "Invalid version number format: " << t.custom_version();
       ver = -1;
     }
-    if (!target_has_tags(t, client_->tags)) {
+    if (!target_has_tags(t, tags)) {
       continue;
     }
     for (const auto& it : t.hardwareIds()) {
       if (it == hwidToFind) {
         targets.emplace_back(t.filename(), t.sha256Hash(), ver, t.custom_data());
         break;
       }
-      for (const auto& hwid : secondary_hwids_) {
+      for (const auto& hwid : secondary_hwids) {
         if (it == Uptane::HardwareIdentifier(hwid)) {
           targets.emplace_back(t.filename(), t.sha256Hash(), ver, t.custom_data());
           break;
         }
       }
     }
   }
-
   std::sort(targets.begin(), targets.end(), compareTargets);
-  return CheckInResult(status, client_->config.provision.primary_ecu_hardware_id, targets);
+  return targets;
+}
+
+CheckInResult AkliteClient::UpdateMetaAndGetTargets(std::shared_ptr<aklite::tuf::RepoSource> repo_src) const {
+  auto status = CheckInResult::Status::Ok;
+
+  LOG_INFO << "Refreshing Targets metadata";
+  try {
+    tuf_repo_->updateMeta(repo_src);
+  } catch (const std::exception& e) {
+    LOG_WARNING << "Unable to update latest metadata, using local copy: " << e.what();
+    try {
+      tuf_repo_->checkMeta();
+    } catch (const std::exception& e) {
+      LOG_ERROR << "Unable to use local copy of TUF data";
+      return CheckInResult(CheckInResult::Status::Failed, "", {});
+    }
+    status = CheckInResult::Status::OkCached;
+  }
+
+  auto allTargetsTuf = tuf_repo_->GetTargets();
+  std::vector<Uptane::Target> allTargets;
+  for (auto const& ut : allTargetsTuf) {
+    allTargets.emplace_back(Target::fromTufTarget(ut));
+  }
+
+  Uptane::HardwareIdentifier hwidToFind(client_->config.provision.primary_ecu_hardware_id);
+  auto matchingTargets = filterTargets(allTargets, hwidToFind, client_->tags, secondary_hwids_);
+  return CheckInResult(status, client_->config.provision.primary_ecu_hardware_id, matchingTargets);
+}
+
+CheckInResult AkliteClient::CheckIn() const {
+  if (!configUploaded_) {
+    client_->reportAktualizrConfiguration();
+    configUploaded_ = true;
+  }
+  client_->reportNetworkInfo();
+  client_->reportHwInfo();
+  client_->reportAppsState();
+
+  boost::property_tree::ptree pt;
+  pt.put<std::string>("tag", client_->tags.empty() ? "" : client_->tags.at(0));
+  auto current = client_->getCurrent();
+  pt.put<std::string>("dockerapps", Target::appsStr(current, ComposeAppManager::Config(client_->config.pacman).apps));
+  pt.put<std::string>("target", current.filename());
+  pt.put<std::string>("ostreehash", current.sha256Hash());
+
+  auto repo_src = std::make_shared<aklite::tuf::AkHttpsRepoSource>("temp-remote-repo-source", pt, client_->config);
```

Comment:
We may wanna optimize it in the future so a new instance of `AkHttpsRepoSource` is not created on each checkin, instead just once when `AkliteClient` is initiated. Actually, this is rather about the http client not `AkHttpRepoSource`, so we may have another ctor for `AkHttpRepoSource` that accepts an http client as an input. Then just the http client is created once in the `AkliteClient` context. As result we'll the TLS/TCP connection can be reused.

It's not critical to do it now, can be done in the following PRs.